src/Compilers/CSharp/Portable/Binder/Imports.cs

```diff
@@ -574,43 +575,83 @@ private void Validate()
+                diagnostics.Clear();
+                diagnostics.AddRange(alias.Alias.AliasTargetDiagnostics);
+
+                alias.Alias.CheckConstraints(diagnostics);
+
+                semanticDiagnostics.AddRange(diagnostics.DiagnosticBag);
+                recordImportDependencies(alias.UsingDirective, target);
             }
 
             var corLibrary = _compilation.SourceAssembly.CorLibrary;
             var conversions = new TypeConversions(corLibrary);
             foreach (var @using in Usings)
             {
+                diagnostics.Clear();
+                diagnostics.AddDependencies(@using.Dependencies);
+
+                NamespaceOrTypeSymbol target = @using.NamespaceOrType;
+
                 // Check if `using static` directives meet constraints.
-                if (@using.NamespaceOrType.IsType)
+                UsingDirectiveSyntax usingDirective = @using.UsingDirective;
+                if (target.IsType)
                 {
-                    var typeSymbol = (TypeSymbol)@using.NamespaceOrType;
-                    var location = @using.UsingDirective?.Name.Location ?? NoLocation.Singleton;
-                    typeSymbol.CheckAllConstraints(_compilation, conversions, location, semanticDiagnostics);
+                    var typeSymbol = (TypeSymbol)target;
+                    var location = usingDirective?.Name.Location ?? NoLocation.Singleton;
+                    typeSymbol.CheckAllConstraints(_compilation, conversions, location, diagnostics);
                 }
+
+                semanticDiagnostics.AddRange(diagnostics.DiagnosticBag);
+                recordImportDependencies(usingDirective, target);
             }
 
             // Force resolution of extern aliases.
             foreach (var alias in ExternAliases)
             {
-                alias.Alias.GetAliasTarget(null);
-                semanticDiagnostics.AddRange(alias.Alias.AliasTargetDiagnostics);
+                var target = (NamespaceSymbol)alias.Alias.GetAliasTarget(null);
+                Debug.Assert(target.IsGlobalNamespace);
+
+                semanticDiagnostics.AddRange(alias.Alias.AliasTargetDiagnostics.DiagnosticBag);
+
+                if (!Compilation.ReportUnusedImportsInTree(alias.ExternAliasDirective.SyntaxTree))
+                {
+                    diagnostics.Clear();
+                    diagnostics.AddAssembliesUsedByNamespaceReference(target);
+                    _compilation.AddUsedAssemblies(diagnostics.DependenciesBag);
+                }
             }
 
             if (_diagnostics != null && !_diagnostics.IsEmptyWithoutResolution)
             {
                 semanticDiagnostics.AddRange(_diagnostics.AsEnumerable());
             }
+
+            diagnostics.Free();
+
+            void recordImportDependencies(UsingDirectiveSyntax usingDirective, NamespaceOrTypeSymbol target)
+            {
+                if (usingDirective is object && Compilation.ReportUnusedImportsInTree(usingDirective.SyntaxTree))
```

Comment:
>ReportUnusedImportsInTree [](start = 60, length = 25)

It feels a little weird that this is a condition being checked in the binder. It seems more like we'd want to record that the dependencies were used in imports, and then it would be the consumer's job to filter them out based on whether they care about imports used or not.

Reply:
> It feels a little weird that this is a condition being checked in the binder. It seems more like we'd want to record that the dependencies were used in imports, and then it would be the consumer's job to filter them out based on whether they care about imports used or not.

The condition doesn't affect the binding, just the dependency management, which is affected by the setting.
This is the last time we look at imports if we don't track their usage and this is the place where we report declaration diagnostics about imports. I do not believe it is worth going through the imports again elsewhere and spending memory just so that we could separate the code. 

---
In reply to: [361555868](https://github.com/dotnet/roslyn/pull/40560#discussion_r361555868) [](ancestors = 361555868)